### PR TITLE
[ci skip] Speed up expand2b benchmark

### PR DESCRIPTION
**Pull request by Sumith1896**
Building onto @certik's works, [here](https://github.com/sympy/symengine/compare/master...certik:int)

The results of various benchmarks can be obtained [here](https://github.com/sympy/sympy/wiki/Benchmark-results-expand2b,-SymEngine)

### CODE REVIEW

**Comment by Sumith1896** [comment]
Implemented packing of `vec_int`.

File: src/dict.cpp [hunk 1]
```diff
@@ -51,6 +51,21 @@ inline std::ostream& print_vec_rcp(std::ostream& out, T& d)
     return out;
 }
 
+unsigned long long pack_vec_int(const vec_int &a) {
+    uint n_var = a.size();
+    uint b = 64 / (n_var + 1) ;
```
Comment (Sumith1896) [inline src/dict.cpp]: I am worried about 'magic' numbers like this, any suggestions for improvement.

**Comment by Sumith1896** [comment]
The current check function `if_pack_vec_int` maybe a bit costly.
The lower bound checks can be removed if we can ensure that we pass a vector of unsigned ints.

**Comment by Sumith1896** [comment]
@certik I think the changes, as and when needed to be merged, can be done in a fresh PR. It'll be cleaner.

**Comment by certik** [comment]
Sure, we can refactor in separate PRs. Now we are just trying to get the speed of  Piranha.

I tested 35f384a484315efbb900ec4ec0b0eb97e791c193. I noticed I have to manually enable MPFR in cmake, otherwise it fails during linking. So one of our dependencies depend on MPFR

Here are my timings on my older laptop:

expand2 1628ms
expand2b 160ms
expand2c 47ms
fateman1_perf 20ms

**Comment by certik** [comment]
Let's see if we can use a faster hashtable. Piranha is using `piranha::hash_set`, which is a custom hash table, similar to `std::unordered_set`. We are very close, we just need to speed it up about 2x, then we'll be fine.

**Comment by Sumith1896** [comment]
MPFR linking is needed in my branch and not in master. Maybe because we are not using Piranha dependency on master. Here the build breaks in `test_arit.cpp` saying wrong reference `mpfr_free_cache`. I have not modified anything there, so I wonder why!

**Comment by isuruf** [comment]
MPFR is a dependency of Piranha, so you need to link with MPFR as well.

**Comment by Sumith1896** [comment]
Okay I'll do that but what surprises me is that it breaks in the area that I have not touched.
Could you reproduce that and let me know why?

**Comment by Sumith1896** [comment]
@certik This PR can be closed now, unless you have anything in mind to speed it up even more.
We should start discussing the design of `Polynomial` now.

**Comment by Sumith1896** [comment]
Also the report is updated, I have couple of queries too.

**Comment by bluescarni** [comment]
@Sumith1896 the MPFR breakage is because Piranha has a global thread pool which, at the exit of each thread, calls ``mpfr_free_cache`` for cleanup purposes (as suggested by the official MPFR documentation). The whole threading machinery is brought into your code by the polynomial multiplication (even if you run all the benchmarks in single-threaded mode).

**Comment by Sumith1896** [comment]
Cool, I'll add MPFR as a Piranha dependency

**Comment by certik** [comment]
@Sumith1896 what is the status of this?

**Comment by Sumith1896** [comment]
@certik I think we have experimented enough, this PR was made so that everybody could try it out.
I'll document the optimizations that we used in the wiki and open a PR for wrappers.
This can be closed.

**Comment by Sumith1896** [comment]
The Polynomial wrappers are being developed here https://github.com/sympy/symengine/pull/511

**Pull request closed without merging**